### PR TITLE
[parametric] feat: add fa2 implementation in typescript

### PR DESCRIPTION
**Pull request by d4hines**
## Depends
- [ ] #650 

## Problem
We need more complex example applications on Deku.

## Solution
Implement the FA2 spec in Typescript.

See documentation: https://tzip.tezosagora.org/proposal/tzip-12/

### CODE REVIEW

File: examples/js-fa2/src/fa2.ts [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+      const json: ledger_entry = JSON.parse(json_str);
+      const next = { ...json, operators: json.operators.filter(op => op !== operator) }
+      deku_storage.set(`${owner}-${token_id}`, JSON.stringify(next));
+    }
+  }
+
+  const get_operators = (address: address, token_id: token_id): address[] => {
+    const json_str = deku_storage.get(`${address}-${token_id}`);
+    if (json_str === undefined) {
+      return []
```
Comment (gsebil08) [inline examples/js-fa2/src/fa2.ts:125]: I think `throw FA2_Error.FA2_TOKEN_UNDEFINED` is a better thing to do.

File: examples/js-fa2/__tests__/fa2-deku-cli.test.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
+import { FA2_Error, update_operator_type } from "../src/fa2";
```
Comment (gsebil08) [inline examples/js-fa2/__tests__/fa2-deku-cli.test.ts:1]: Global TS question, do you use a particular TS formatter?

**Review by gsebil08 — COMMENTED** [review]
I am not relevant to review the real subject handled by this PR... But I tried...

File: examples/js-fa2/src/fa2.ts (hunk 1 above)
Comment (Pilou97) [reply examples/js-fa2/src/fa2.ts:125]: In some case we don't want to throw exception, in this case the default value should be an empty array, you can see the [tzip-12 implementations](https://gitlab.com/tezos/tzip/-/blob/master/proposals/tzip-12/tzip-12.md) : 
> It is possible to update operators for a token owner that does not hold any token
balances yet.

So even if there isn't a token for a given user, you can interpret its operators as an empty array

File: examples/js-fa2/src/fa2.ts [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+  FA2_NOT_OWNER = "fa2_not_owner", // A transfer failed because operator_transfer_policy == Owner_transfer and it is invoked not by the token owner
+  FA2_NOT_OPERATOR = "fa2_not_operator", // A transfer failed because operator_transfer_policy == Owner_or_operator_transfer and it is invoked neither by the token owner nor a permitted operator
+  FA2_OPERATORS_UNSUPPORTED = "fa2_operators_unsupported", // update_operators entrypoint is invoked and operator_transfer_policy is No_transfer or Owner_transfer
+  FA2_RECEIVER_HOOK_FAILED = "fa2_receiver_hook_failed", // The receiver hook failed. This error MUST be raised by the hook implementation
+  FA2_SENDER_HOOK_FAILED = "fa2_sender_hook_failed", // The sender failed. This error MUST be raised by the hook implementation
+  FA2_RECEIVER_HOOK_UNDEFINED = "fa2_receiver_hook_undefined", // Receiver hook is required by the permission behavior, but is not implemented by a receiver contract
+  FA2_SENDER_HOOK_UNDEFINED = "fa2_sender_hook_undefined", // Sender hook is required by the permission behavior, but is not implemented by a sender contract
+}
+
+export enum FA2_Custom_Error {
```
[the remark this answers is not in the document]
Comment (Pilou97) [reply examples/js-fa2/src/fa2.ts:34]: I wanted to differentiate errors between the needed ones by the tzip-12 and the custom errors

File: examples/js-fa2/src/fa2.ts [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+  get_operators: (address: address, token_id: token_id) => address[];
+  // returns the metadata on a token
+  get_token_metadata: (token_id: token_id) => token_metadata
+  // returns the administrator of the fa2 storage
+  get_administrator: () => address | undefined
+  // add a token in the store
+  add_token: (token_id: token_id, token_metadata: token_metadata, amount: amount) => void;
+}
+
+// This function is used to create a fa2_storage (see the interface) form the Deku_storage
```
[the remark this answers is not in the document]
Comment (Pilou97) [reply examples/js-fa2/src/fa2.ts:66]: In deku parametric, there is a state, a map keys<string> values<json>, and from the state machine you can set and get thoses values

File: examples/js-fa2/src/fa2.ts [hunk 5]
```diff
@@ -0,0 +1,376 @@
+"use strict";
+
+type nat = number; // TODO: create a real constructor to avoid negative number
```
[the remark this answers is not in the document]
Comment (d4hines) [reply examples/js-fa2/src/fa2.ts:3]: A `TODO:` is enough here. This application doesn't have to be perfect.

**Pull request closed without merging**